course/src/isometry.cc

```diff
@@ -0,0 +1,305 @@
+    r1_[1] = *it++;
+    r1_[2] = *it++;
+
+    r2_[0] = *it++;
+    r2_[1] = *it++;
+    r2_[2] = *it++;
+
+    r3_[0] = *it++;
+    r3_[1] = *it++;
+    r3_[2] = *it++;
+}
+
+Matrix3::Matrix3(const std::initializer_list<double>& l1,
+                 const std::initializer_list<double>& l2,
+                 const std::initializer_list<double>& l3) {
+    r1_ = Vector3(l1);
+    r2_ = Vector3(l2);
+    r3_ = Vector3(l3);
+}
+
+// Operators
+Vector3& Matrix3::operator [] (const uint8_t index) {
+    switch(index) {
+        case 0:
+            return r1_;
+        case 1:
+            return r2_;
+        case 2:
+            return r3_;
+        default:
+            throw "Error. Invalid row index for Matrix3";
+    }
+}
+
+const Vector3& Matrix3::operator [] (const uint8_t index) const {
+    switch(index) {
+        case 0:
+            return r1_;
+        case 1:
+            return r2_;
+        case 2:
+            return r3_;
+        default:
+            throw "Error. Invalid row index for Matrix3";
+    }
+}
+
+Matrix3 Matrix3::operator + (const Matrix3& m) const {
+    return Matrix3(r1_ + m.r1_, r2_ + m.r2_, r3_ + m.r3_);
+}
+
+Matrix3 Matrix3::operator - (const Matrix3& m) const {
+    return Matrix3(r1_ - m.r1_, r2_ - m.r2_, r3_ - m.r3_);
+}
+
+Matrix3 Matrix3::operator * (const Matrix3& m) const {
+    return Matrix3(r1_ * m.r1_, r2_ * m.r2_, r3_ * m.r3_);
+}
+
+Matrix3 Matrix3::operator * (const double d) const {
+    return Matrix3(r1_ * d, r2_ * d, r3_ * d);
+}
+
+Matrix3 Matrix3::operator / (const Matrix3& m) const {
+    return Matrix3(r1_ / m.r1_, r2_ / m.r2_, r3_ / m.r3_);
+}
+
+Matrix3& Matrix3::operator = (const Matrix3& m) {
+    r1_ = m.r1_;
+    r2_ = m.r2_;
+    r3_ = m.r3_;
+    return *this;
+}
+
+Matrix3& Matrix3::operator += (const Matrix3& m) {
+    r1_ += m.r1_;
+    r2_ += m.r2_;
+    r3_ += m.r3_;
+    return *this;
+}
+
+Matrix3& Matrix3::operator -= (const Matrix3& m) {
+    r1_ -= m.r1_;
+    r2_ -= m.r2_;
+    r3_ -= m.r3_;
+    return *this;
+}
+
+Matrix3& Matrix3::operator *= (const Matrix3& m) {
+    r1_ *= m.r1_;
+    r2_ *= m.r2_;
+    r3_ *= m.r3_;
+    return *this;
+}
+
+Matrix3& Matrix3::operator /= (const Matrix3& m) {
+    r1_ /= m.r1_;
+    r2_ /= m.r2_;
+    r3_ /= m.r3_;
+    return *this;
+}
+
+bool Matrix3::operator == (const Matrix3& m) {
+    if ((r1_ == m.r1_) && (r2_ == m.r2_) && (r3_ == m.r3_)) {
+        return true;
+    }
+    return false;
+}
+
+bool Matrix3::operator != (const Matrix3& m) {
+    return !(*this == m);
+}
+
+Vector3 Matrix3::operator * (const Vector3& v) {
+    return Vector3(r1_.dot(v), r2_.dot(v), r3_.dot(v));
+}
+
+Matrix3& Matrix3::operator = (Matrix3&& m) {
+    r1_ = std::move(m.r1_);
+    r2_ = std::move(m.r2_);
+    r3_ = std::move(m.r3_);
+
+    return *this; // What about the original matrix? (m) It shouldn't have its rows available, if I'm not mistaken
+}
+
+// Getters
+Vector3 Matrix3::row(uint8_t index) const {
+    return Vector3((*this)[index]);
```

Comment:
I don't think you need to create a copy of `Vector3`. `return (*this)[index];` should be enough.